src/kde/kde.cc

```diff
@@ -30,124 +30,437 @@ namespace rtt_kde {
       }
     }
   } else { // local reconstruction only
+
+    // if the log bias is zero the answer must be zero everywhere
+    if (!(log_bias > 0.0))
+      return result;
+
     // now apply the kernel to the local ranks
-    for (int i = 0; i < local_size; i++) {
-      const double x0 = position[i][0];
-      const double one_over_h = one_over_band_width[i][0];
+    std::array<double, 3> win_min{0.0, 0.0, 0.0};
+    std::array<double, 3> win_max{0.0, 0.0, 0.0};
+    for (size_t i = 0; i < local_size; i++) {
+      const std::array<double, 3> r0 = qindex.locations[i];
+      const std::array<double, 3> one_over_h0 = one_over_bandwidth[i];
+      calc_win_min_max(qindex, r0, one_over_h0, win_min, win_max);
+      const std::vector<size_t> coarse_bins = qindex.window_coarse_index_list(win_min, win_max);
       // fetch local contribution
-      for (int j = 0; j < local_size; j++) {
-        const double x = position[j][0];
-        const double u = (x0 - x) * one_over_h;
-        const double weight = (epan_kernel(u)) * one_over_h;
-        result[i] += distribution[j] * weight;
-        normal[i] += weight;
+      for (auto &cb : coarse_bins) {
+        // skip bins that aren't present in the map (can't use [] operator with constness)
+        auto mapItr = qindex.coarse_index_map.find(cb);
+        if (mapItr != qindex.coarse_index_map.end()) {
+          // loop over local data
+          for (auto &l : mapItr->second) {
+            const double weight = calc_weight(r0, one_over_h0, qindex.locations[l],
+                                              one_over_bandwidth[l], qindex, discontinuity_cutoff);
+            result[i] += log_transform(distribution[l], log_bias) * weight;
+            normal[i] += weight;
+          }
+        }
       }
     }
   }
 
   // normalize the integrated weight contributions
-  for (int i = 0; i < local_size; i++) {
+  for (size_t i = 0; i < local_size; i++) {
     Check(normal[i] > 0.0);
     result[i] /= normal[i];
+    result[i] = log_inv_transform(result[i], log_bias);
+    // ZERO IS ZERO AND THE LOG TRANSFORM CAN MAKE THE ZEROS NOT MATCH... SO FIX IT LIKE THIS
+    if (rtt_dsxx::soft_equiv(result[i], 0.0) && rtt_dsxx::soft_equiv(distribution[i], 0.0))
+      result[i] = distribution[i];
   }
 
-  double reconstruction_conservation = std::accumulate(result.begin(), result.end(), 0.0);
+  return result;
+}
+
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief KDE apply conservation
+ * 
+ * \pre Apply conservation fix to the new distribution so
+ * sum(original_distribution) == sum(new_distribution)
+ *
+ * \param[in] original_distribution original data to be reconstructed
+ * \param[in,out] new_distribution original data to be reconstructed
+ * \param[in] domain_decomposed bool
+ *
+ */
+void kde::apply_conservation(const std::vector<double> &original_distribution,
+                             std::vector<double> &new_distribution,
```

Comment:
Can you return `new_distribution` instead of passing a reference? Or is this already allocated?